File: thrifts-java/thrifts-serializer/src/main/java/thrifts/serializer/TProtocolExtension.java
package thrifts.serializer;

import com.fasterxml.classmate.ResolvedType;
import com.fasterxml.classmate.TypeResolver;
import org.apache.thrift.TException;
import org.apache.thrift.protocol.*;
import thrifts.common.ThriftSException;
import thrifts.common.annotations.ThriftSMember;

import java.lang.reflect.Field;
import java.lang.reflect.Type;
import java.nio.ByteBuffer;
import java.util.*;

final class TProtocolExtension {
    static final TypeResolver typeResolver = new TypeResolver();

    public static void writeStruct(TProtocol protocol, TStruct struct, Object instance) throws TException, IllegalAccessException, ThriftSException {
        protocol.writeStructBegin(struct);

        Class<?> type = instance.getClass();
        Field[] props = type.getDeclaredFields();
        for (Field prop : props) {
            short fieldId = 0;
            ThriftSMember attr = prop.getAnnotation(ThriftSMember.class);
            if (attr != null) {
                fieldId = attr.tag();
                prop.setAccessible(true);
                writeField(protocol, fieldId, prop.getName(), prop.get(instance));
            }
        }

        protocol.writeFieldStop();
        protocol.writeStructEnd();
    }

    private static void writeField(TProtocol protocol, short fieldId, String fieldName, Object value) throws TException, ThriftSException, IllegalAccessException {
        if (value == null) return;

        //java类型擦除
        Class<?> sourceType = value.getClass();

        byte fieldType = TypeExtension.toThriftType(typeResolver.resolve(sourceType));
        TField field = new TField(fieldName, fieldType, fieldId);


        protocol.writeFieldBegin(field);
        writeFieldValue(protocol, sourceType, field.type, value);
        protocol.writeFieldEnd();
    }

    private static void writeFieldValue(TProtocol protocol, Class<?> sourceType, byte thriftType, Object value) throws TException, IllegalAccessException, ThriftSException {
        switch (thriftType) {
            case TType.BOOL:
                protocol.writeBool(((Boolean) value).booleanValue());
                break;
            case TType.BYTE:
                protocol.writeByte(((Byte) value).byteValue());
                break;
            case TType.I16:
                protocol.writeI16(((Short) value).shortValue());
                break;
            case TType.I32:
                protocol.writeI32(((Integer) value).intValue());
                break;
            case TType.I64:
                if (sourceType == Date.class) {
//                    protocol.writeI64(DateTimeHelper.ToUnixTimestamp(Convert.ToDateTime(value)));
                } else {
                    protocol.writeI64(((Long) (value)).longValue());
                }

                break;
            case TType.DOUBLE:
                protocol.writeDouble(((Double) (value)).doubleValue());
                break;
            case TType.STRING:
                if (sourceType == String.class
                        || sourceType == UUID.class) {
                    protocol.writeString(value.toString());
                } else if (sourceType == byte[].class) {
                    protocol.writeBinary(ByteBuffer.wrap((byte[]) value));
                }
                break;
            case TType.STRUCT:
                writeStruct(protocol, new TStruct(), value);
                break;
            case TType.LIST:
                List<?> list = (List<?>) value;

                byte itemType = TType.STRUCT;
                if(list.size()>0)
                {
                    itemType = TypeExtension.toThriftType(typeResolver.resolve(list.get(0).getClass()));
                }
                protocol.writeListBegin(new TList(TType.STRUCT, list.size()));
                for (Object item : list) {
                    //排除后和count不一致
                    if (item == null) {
                        throw new ThriftSException("The item must be not null.");
                    }

                    writeFieldValue(protocol, item.getClass(), itemType, item);
                }
                protocol.writeListEnd();
                break;
            case TType.MAP:
                Map<?, ?> map = (Map<?, ?>) value;
//                Type[] genericArgs = ((ParameterizedType) value.getClass().getGenericSuperclass()).getActualTypeArguments();
                byte keyType = TType.STRUCT;
                byte valueType = TType.STRUCT;
                if(map.size()>0)
                {
                    keyType = TypeExtension.toThriftType(typeResolver.resolve(map.keySet().toArray()[0].getClass()));
                    valueType = TypeExtension.toThriftType(typeResolver.resolve(map.values().toArray()[0].getClass()));
                }

                protocol.writeMapBegin(new TMap(
                        keyType,
                        valueType,
                        map.size()));
//                protocol.writeMapBegin(new TMap(
//                        TypeExtension.toThriftType((Class<?>) genericArgs[0]),
//                        TypeExtension.toThriftType((Class<?>) genericArgs[1]),
//                        map.size()));
                for (Object key : map.keySet()) {
                    //排除后和count不一致
                    if (key == null || map.get(key) == null) {
                        throw new ThriftSException("The key or value must be not null.");
                    }
                    writeFieldValue(protocol,
                            key.getClass(),
                            TypeExtension.toThriftType(typeResolver.resolve(key.getClass())),
                            key);

                    writeFieldValue(protocol,
                            map.get(key).getClass(),
                            TypeExtension.toThriftType(typeResolver.resolve(map.get(key).getClass())),
                            map.get(key));
                }
                protocol.writeMapEnd();
                break;
            default:
                break;
        }
    }

    public static Object readStruct(TProtocol protocol, ResolvedType type) throws IllegalAccessException, InstantiationException, TException {
        Object result = null;
        HashMap<Short, Field> dicProps = new HashMap<Short, Field>();

//        Class<?> targetType = null;
//        if(type instanceof ParameterizedType) {
//            targetType = (Class<?>) ((ParameterizedType) type).getRawType();
//        }
//        else {
//            targetType = (Class<?>) type;
//        }

        Class<?> targetType = type.getErasedType();

        result = targetType.newInstance();

        if (targetType.isPrimitive() == false)
        {
            Field[] props = targetType.getDeclaredFields();
            for(Field prop : props)
            {
                ThriftSMember attr = prop.getAnnotation(ThriftSMember.class);
                if(attr!=null){
                    dicProps.put(attr.tag(), prop);
                }
            }
        }

        TStruct struct = protocol.readStructBegin();

        while (true)
        {
            TField field = protocol.readFieldBegin();
            if (field.type == TType.STOP)
            {
                break;
            }

            // 跳过不存在的字段
            if (dicProps.containsKey(field.id))
            {
                Object fieldValue = null;

                Field prop = dicProps.get(field.id);
                prop.setAccessible(true);

                //fieldValue = readFieldValue(protocol, field.type, prop.getGenericType());

                List<ResolvedType> typeParam = type.getTypeParameters();


                if(result instanceof ValueContainer)
                {
                    ResolvedType tt = typeResolver.resolve(typeParam.get(0).getErasedType());
                    fieldValue = readFieldValue(protocol, field.type, tt);
                }
                else if(result instanceof ListContainer)
                {
                    ResolvedType tt = typeResolver.resolve(ArrayList.class,typeParam.get(0).getErasedType());
                    fieldValue = readFieldValue(protocol, field.type, tt);
                }
                else if(result instanceof MapContainer)
                {
                    ResolvedType tt = typeResolver.resolve(HashMap.class,typeParam.get(0),typeParam.get(1));
                    fieldValue = readFieldValue(protocol, field.type, tt);
                }
                else
                {
                    fieldValue = readFieldValue(protocol, field.type, typeResolver.resolve(prop.getGenericType()));
                }

                if (prop.getDeclaringClass() == UUID.class)
                {
                    prop.set(result, UUID.fromString(fieldValue.toString()));
                }
                else if (prop.getDeclaringClass() == Date.class)
                {
//                    prop.set(result, DateTimeHelper.ToDateTime((long) fieldValue));
                }
                else if (prop.getDeclaringClass().isEnum())
                {
                    Object enumValue = Enum.valueOf((Class<Enum>) prop.getDeclaringClass(), fieldValue.toString());
                    prop.set(result, enumValue);
                }
                else if (prop.getDeclaringClass().isPrimitive()
                        && prop.getDeclaringClass() != fieldValue.getClass())
                {
                    // ex: double --> decimal
//                    var newFieldValue = Convert.ChangeType(fieldValue, dicProps[field.ID].PropertyType);
//                    dicProps.get(field.id).set(result, newFieldValue);
                }
                else
                {
                    prop.set(result, fieldValue);
                }
            }

            protocol.readFieldEnd();
        }

        protocol.readStructEnd();
        return result;
    }

    private static Object readFieldValue(TProtocol protocol, byte thriftType, ResolvedType type) throws IllegalAccessException, TException, InstantiationException {
        Class<?> targetType = null;
        Type[] genericArguments = null;
//        TypeResolver typeResolver = new TypeResolver();
//        if(type instanceof ParameterizedType)
//        {
//            targetType = (Class<?>)((ParameterizedType)type).getRawType();
//            genericArguments = ((ParameterizedType)type).getActualTypeArguments();
//        }
//        else
//        {
//            targetType = (Class<?>)type.getErasedType();
//        }
        targetType = (Class<?>)type.getErasedType();

        switch (thriftType)
        {
            //case TType.Void:
            //    return null;
            case TType.BOOL:
                return protocol.readBool();
            case TType.BYTE:
                return protocol.readByte();
            case TType.I16:
                return protocol.readI16();
            case TType.I32:
                return protocol.readI32();
            case TType.I64:
                return protocol.readI64();
            case TType.DOUBLE:
                return protocol.readDouble();
            case TType.STRING:
                if (targetType == String.class
                        || targetType == UUID.class)
                {
                    return protocol.readString();
                }
                else
                {
                    return protocol.readBinary();
                }
            case TType.STRUCT:
                return readStruct(protocol, type);
            case TType.LIST:
                Object result = targetType.newInstance();
                ArrayList list = (ArrayList)result;

                // 只考虑泛型list
//                if (targetType.IsGenericType)
//                {
//                    Class<?> genericType = (Class<?>)((ParameterizedType) targetType.getGenericSuperclass()).getActualTypeArguments()[0];
                    //Class<?> genericType = (Class<?>)genericArguments[0];
                ResolvedType genericType = type.getTypeParameters().get(0);
                    TList tlist = protocol.readListBegin();
                    for (int i = 0; i < tlist.size; i++)
                    {
                        Object item = readFieldValue(protocol, TypeExtension.toThriftType(genericType), genericType);
                        list.add(item);
                    }
                    protocol.readListEnd();
//                }

                return result;

            case TType.MAP:
                Object resultMap = targetType.newInstance();
                //Map<?,?> dic = (Map<?,?>)resultMap;
                Map<Object,Object> dic = (Map<Object,Object>)resultMap;

                // 只考虑泛型map
//                if (targetType.IsGenericType)
//                {
//                    Type[] genericTypes = ((ParameterizedType) targetType.getGenericSuperclass()).getActualTypeArguments();
//                    Class<?> genericTypeKey = (Class<?>)genericTypes[0];
//                    Class<?> genericTypeValue = (Class<?>)genericTypes[1];

                    Class<?> genericTypeKey = type.getTypeParameters().get(0).getErasedType(); //(Class<?>)genericArguments[0];

                ResolvedType valType = null;
                    if(type.getTypeParameters().get(1).getTypeParameters().size()>0)
                    {
                        valType = type.getTypeParameters().get(1);
                    }
                    else
                    {
                        Class<?> genericTypeValue = type.getTypeParameters().get(1).getErasedType();;
                        valType = typeResolver.resolve(genericTypeValue);
                    }

//                    Type[] genericArgumentsNext = null;
//                    if(genericArguments[1] instanceof ParameterizedType)
//                    {
//                        ParameterizedType genericTypeValueP = (ParameterizedType)genericArguments[1];
//                        genericTypeValue = (Class<?>)genericTypeValueP.getRawType();
//                        genericArgumentsNext = genericTypeValueP.getActualTypeArguments();
//                    }
//                    else
//                    {
//                        genericTypeValue = (Class<?>)genericArguments[1];
//                    }


                    TMap tmap = protocol.readMapBegin();
                    for (int i = 0; i < tmap.size; i++)
                    {
                        Object key = readFieldValue(protocol, TypeExtension.toThriftType(typeResolver.resolve(genericTypeKey)), typeResolver.resolve(genericTypeKey));
                        Object value = readFieldValue(protocol, TypeExtension.toThriftType(valType),valType );
                        dic.put(key, value);
                    }
                    protocol.readMapEnd();
//                }

                return resultMap;
            default:
                TProtocolUtil.skip(protocol, thriftType);
                return null;
        }
    }
}

File: thrifts-dotnet/ThriftS.Serializer/TProtocolExtension.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Thrift.Protocol;
using ThriftS.Common;
using ThriftS.Common.Attributes;

namespace ThriftS.Serializer
{
    internal static class TProtocolExtension
    {
        public static void WriteStruct(this TProtocol protocol, TStruct @struct, object instance)
        {
            protocol.WriteStructBegin(@struct);

            var type = instance.GetType();
            var props = type.GetProperties();
            foreach (var prop in props)
            {
                short fieldId = 0;
                var attr = prop.GetCustomAttributes(typeof(ThriftSMemberAttribute), false).FirstOrDefault();
                if (attr != null)
                {
                    fieldId = (attr as ThriftSMemberAttribute).Tag;
                    protocol.WriteField(fieldId, prop.Name, prop.PropertyType, prop.GetValue(instance));
                }
            }

            protocol.WriteFieldStop();
            protocol.WriteStructEnd();
        }

        private static void WriteField(this TProtocol protocol, short fieldId, string fieldName, Type sourceType,
            object value)
        {
            if (value == null) return;

            var field = new TField();
            field.ID = fieldId;
            field.Name = fieldName;
            field.Type = sourceType.ToThriftType();

            protocol.WriteFieldBegin(field);
            protocol.WriteFieldValue(sourceType, field.Type, value);
            protocol.WriteFieldEnd();
        }

        private static void WriteFieldValue(this TProtocol protocol, Type sourceType, TType thriftType, object value)
        {
            switch (thriftType)
            {
                case TType.Bool:
                    protocol.WriteBool(Convert.ToBoolean(value));
                    break;
                case TType.Byte:
                    protocol.WriteByte(Convert.ToSByte(value));
                    break;
                case TType.I16:
                    protocol.WriteI16(Convert.ToInt16(value));
                    break;
                case TType.I32:
                    protocol.WriteI32(Convert.ToInt32(value));
                    break;
                case TType.I64:
                    if (sourceType == typeof(DateTime)|| sourceType == typeof(DateTime?))
                    {
                        protocol.WriteI64(DateTimeHelper.ToUnixTimestamp(Convert.ToDateTime(value)));
                    }
                    else
                    {
                        protocol.WriteI64(Convert.ToInt64(value));
                    }

                    break;
                case TType.Double:
                    protocol.WriteDouble(Convert.ToDouble(value));
                    break;
                case TType.String:
                    if (sourceType == typeof(string)
                        || sourceType == typeof(Guid))
                    {
                        protocol.WriteString(Convert.ToString(value));
                    }
                    else if (sourceType == typeof(byte[]))
                    {
                        protocol.WriteBinary((byte[])value);
                    }
                    break;
                case TType.Struct:
                    protocol.WriteStruct(new TStruct(), value);
                    break;
                case TType.List:
                    var list = value as IList;
                    protocol.WriteListBegin(new TList(TType.Struct, list.Count));
                    foreach (var item in list)
                    {
                        //排除后和count不一致
                        if (item == null)
                        {
                            throw new ThriftSException("The item must be not null.");
                        }

                        protocol.WriteFieldValue(item.GetType(), item.GetType().ToThriftType(), item);
                    }
                    protocol.WriteListEnd();
                    break;
                case TType.Map:
                    var map = value as IDictionary;
                    var genericArgs = value.GetType().GetGenericArguments();
                    protocol.WriteMapBegin(new TMap(genericArgs[0].ToThriftType(), genericArgs[1].ToThriftType(), map.Count));
                    foreach (var key in map.Keys)
                    {
                        //排除后和count不一致
                        if (key == null || map[key] == null)
                        {
                            throw new ThriftSException("The key or value must be not null.");
                        }
                        protocol.WriteFieldValue(key.GetType(), key.GetType().ToThriftType(), key);
                        protocol.WriteFieldValue(map[key].GetType(), map[key].GetType().ToThriftType(), map[key]);
                    }
                    protocol.WriteMapEnd();
                    break;
                default:
                    break;
            }
        }

        public static object ReadStruct(this TProtocol protocol, Type targetType)
        {
            object result = null;
            var dicProps = new Dictionary<int, PropertyInfo>();

            result = Activator.CreateInstance(targetType);

            if (targetType.IsClass)
            {
                var props = targetType.GetProperties();
                foreach (var prop in props)
                {
                    var attrs = prop.GetCustomAttributes(typeof(ThriftSMemberAttribute), false);
                    if (attrs.Any())
                    {
                        dicProps.Add((attrs[0] as ThriftSMemberAttribute).Tag, prop);
                    }
                }
            }

            var @struct = protocol.ReadStructBegin();

            while (true)
            {
                var field = protocol.ReadFieldBegin();
                if (field.Type == TType.Stop)
                {
                    break;
                }

                // 跳过不存在的字段
                if (dicProps.ContainsKey(field.ID))
                {
                    var fieldValue = protocol.ReadFieldValue(field.Type, dicProps[field.ID].PropertyType);

                    if (dicProps[field.ID].PropertyType == typeof(Guid))
                    {
                        dicProps[field.ID].SetValue(result, new Guid(fieldValue.ToString()));
                    }
                    else if (dicProps[field.ID].PropertyType == typeof(DateTime)||dicProps[field.ID].PropertyType == typeof(DateTime?))
                    {
                        dicProps[field.ID].SetValue(result, DateTimeHelper.ToDateTime((long)fieldValue));
                    }
                    else if (dicProps[field.ID].PropertyType.IsEnum)
                    {
                        var enumValue = Enum.Parse(dicProps[field.ID].PropertyType, fieldValue.ToString());
                        dicProps[field.ID].SetValue(result, enumValue);
                    }
                    else if (dicProps[field.ID].PropertyType.IsValueType
                             && dicProps[field.ID].PropertyType != fieldValue.GetType())
                    {
                        // ex: double --> decimal
                        var newFieldValue = Convert.ChangeType(fieldValue, dicProps[field.ID].PropertyType);
                        dicProps[field.ID].SetValue(result, newFieldValue);
                    }
                    else
                    {
                        dicProps[field.ID].SetValue(result, fieldValue);
                    }
                }

                protocol.ReadFieldEnd();
            }

            protocol.ReadStructEnd();
            return result;
        }

        private static object ReadFieldValue(this TProtocol protocol, TType thriftType, Type targetType)
        {
            switch (thriftType)
            {
                //case TType.Void:
                //    return null;
                case TType.Bool:
                    return protocol.ReadBool();
                case TType.Byte:
                    return protocol.ReadByte();
                case TType.I16:
                    return protocol.ReadI16();
                case TType.I32:
                    return protocol.ReadI32();
                case TType.I64:
                    return protocol.ReadI64();
                case TType.Double:
                    return protocol.ReadDouble();
                case TType.String:
                    if (targetType == typeof(string)
                        || targetType == typeof(Guid))
                    {
                        return protocol.ReadString();
                    }
                    else
                    {
                        return protocol.ReadBinary();
                    }
                case TType.Struct:
                    return protocol.ReadStruct(targetType);
                case TType.List:
                    var result = Activator.CreateInstance(targetType);
                    var list = result as IList;

                    // 只考虑泛型list
                    if (targetType.IsGenericType)
                    {
                        var genericType = targetType.GetGenericArguments()[0];
                        var tlist = protocol.ReadListBegin();
                        for (int i = 0; i < tlist.Count; i++)
                        {
                            var item = protocol.ReadFieldValue(genericType.ToThriftType(), genericType);
                            list.Add(item);
                        }
                        protocol.ReadListEnd();
                    }

                    return result;

                case TType.Map:
                    var resultMap = Activator.CreateInstance(targetType);
                    var dic = resultMap as IDictionary;

                    // 只考虑泛型map
                    if (targetType.IsGenericType)
                    {
                        var genericTypes = targetType.GetGenericArguments();
                        var genericTypeKey = genericTypes[0];
                        var genericTypeValue = genericTypes[1];

                        var tmap = protocol.ReadMapBegin();
                        for (int i = 0; i < tmap.Count; i++)
                        {
                            var key = protocol.ReadFieldValue(genericTypeKey.ToThriftType(), genericTypeKey);
                            var value = protocol.ReadFieldValue(genericTypeValue.ToThriftType(), genericTypeValue);
                            dic.Add(key, value);
                        }
                        protocol.ReadMapEnd();
                    }

                    return resultMap;
                default:
                    TProtocolUtil.Skip(protocol, thriftType);
                    return null;
            }
        }
    }
}
